pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DateTimeFunctions.java

```diff
@@ -704,4 +707,24 @@ public static long timestampDiff(String unit, long timestamp1, long timestamp2)
     ISOChronology chronology = ISOChronology.getInstanceUTC();
     return DateTimeUtils.getTimestampField(chronology, unit).getDifferenceAsLong(timestamp2, timestamp1);
   }
+
+  /**
+   * For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time,
+   * traditionally known as Greenwich Mean Time, GMT). This method will transform that timestamp to a timestamp
+   * that represents the millis since epoch in the specified time zone.
+   *
+   * @return a timestamp that corresponds to the {@code timestampInUTC} at {@code timeZone}
+   */
+  @ScalarFunction
+  public static Timestamp atTimeZone(long timestampInUTC, String timeZone) {
```

Comment:
High level question: millis since epoch should be independent of the time zone. I don't think there is a way to represent the time zone within the `Timestamp`. To change the time zone, we will need to add the `DateTime` type support. For now as a work around, we can potentially use `String` to represent the value

Reply:
I agree it's a bit tricky - in Postgres it returns a `Timestamp WITH TIME ZONE` type, which maybe we need to add to make this work seamlessly. You can see the following example:
```sql
CREATE TABLE foo (ts TIMESTAMP);
INSERT INTO foo (ts) VALUES (TIMESTAMP '2022-10-23 09:00:00')

--- you can see that the default behavior returns what we stored
--- which is, by default, UTC time
SELECT ts FROM FOO;
ts
------------------------
 2022-10-23T09:00:00Z
------------------------

--- if I ask for it "interpreted as a PST time" then I'll get
--- the UTC timestamp for the corresponding datetime in PST
SELECT ts AT TIME ZONE 'PST' FROM foo;
ts
------------------------
2022-10-23T17:00:00Z
------------------------

--- this shows the degenerate case, when I say that the original
--- timestamp I stored is in 'UTC', but I actually want the timestamp
--- returned to me with what that corresponds to in 'PST'
SELECT ts AT TIME ZONE 'UTC' AT TIME ZONE 'PST' FROM foo;
ts
------------------------
2022-10-23T01:00:00Z
------------------------
```

It's all very confusing!

I thought about whether having this type or not is a problem, and I'm not sure that it is if we always store the timestamp as `epochMillis`. The only difference is that when we introduce `AT TIME ZONE` types, we would just need to convert the type _before_ we store anything (as opposed to assuming that the timestamp that we are given is in UTC). See the difference in behavior for Postgres:
```sql
CREATE TABLE foo (ts TIMESTAMP);
INSERT INTO foo (ts) VALUES (TIMESTAMP '2022-10-23 09:00:00' AT TIME ZONE 'PST')
SELECT ts FROM foo;
ts
------------------------
2022-10-23T17:00:00Z
-----------------------
```
vs
```sql
CREATE TABLE foo (ts TIMESTAMP WITH TIME ZONE);
INSERT INTO foo (ts) VALUES (TIMESTAMP '2022-10-23 09:00:00' AT TIME ZONE 'PST')
SELECT ts AT TIME ZONE 'PST' FROM foo;
ts
-----------------------
2022-10-23T09:00:00Z
-----------------------
```

I'll make sure to double check that all the behavior with this PR is as expected, but I think it should be.

(FWIW, in SQL the `DateTime` type is generally considered identical to `Timestamp` - https://www.postgresql.org/docs/current/datatype-datetime.html)

Reply:
Ideally we want to introduce the `TIMESTAMP WITH TIME ZONE` type, so that it can be added seamlessly. If this feature is urgent, we can add it as is for now, and then change it when the new type support is added. One problem I can see from using `TIMESTAMP` is that comparison logic will break when comparing it from different time zone.

The postgres result is actually confusing to me because the `Z` should represent time zone `UTC`, but it is actually returning the timestamp in `PST`

Reply:
I agree with you - I'll double check with the user that requested this and if it isn't super urgent we can work on adding the type first. Otherwise I'll change it to String return type.

> The postgres result is actually confusing to me because the Z should represent time zone UTC, but it is actually returning the timestamp in PST

Now that you mention it, that is weird...

Reply:
joda.DateTime returns a Timestamp/Epoch long value that's 8hrs away from UTC if the DateTime set timezone to PST.

Reply:
Confirmed Postgres has the same behavior:
```
SELECT 
  extract(epoch from ts AT TIME ZONE 'UTC'), 
  extract(epoch from ts AT TIME ZONE 'PST') 
 FROM foo;
```
returns
```
date_part  | date_part
---------- | ----------
1666515600 | 1666544400


```